compiler/rustc_codegen_ssa/src/base.rs

```diff
@@ -429,9 +429,11 @@ pub fn maybe_create_entry_wrapper<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
         rust_main_def_id: DefId,
         entry_type: EntryFnType,
     ) -> Bx::Function {
-        // The entry function is either `int main(void)` or `int main(int argc, char **argv)`,
-        // depending on whether the target needs `argc` and `argv` to be passed in.
-        let llfty = if cx.sess().target.main_needs_argc_argv {
+        // The entry function is either `int main(void)` or `int main(int argc, char **argv)`, or
+        // `Status efi_main(Handle hd, SystemTable *st)` depending on the target.
+        let llfty = if cx.sess().target.os.contains("uefi") {
+            cx.type_func(&[cx.type_i8p(), cx.type_i8p()], cx.type_isize())
```

Comment:
Is `type_i8p` really correct? Shouldn't this be a void-pointer? I am unsure on what level this operates, so char-pointer might be fine.

Reply:
Well, the only pointer type exposed by rust llvm bindings was `i8p`, at least when I started, will check again. Also, I think from llvm 15, every pointer is a void pointer, but I will have to check that.

Reply:
Checking the generated llvm IR for `x86_64-unknown-uefi`:
```llvm-ir
; Function Attrs: noredzone nounwind
define win64cc i64 @efi_main(ptr %0, ptr %1) unnamed_addr #14 {
top:
  %_9.i = alloca ptr, align 8
  %2 = alloca [2 x ptr], align 8
  store ptr %0, ptr %2, align 8
  %3 = getelementptr inbounds ptr, ptr %2, i64 1
  store ptr %1, ptr %3, align 8
  call void @llvm.lifetime.start.p0(i64 8, ptr nonnull %_9.i)
  store ptr @_ZN11hello_world4main17hb0026451eb24fd28E, ptr %_9.i, align 8
; call std::rt::lang_start_internal
  %4 = call i64 @_ZN3std2rt19lang_start_internal17h2e5a760b138f1504E(ptr noundef nonnull align 1 %_9.i, ptr noalias noundef nonnull readonly align 8 dereferenceable(24) @vtable.0, i64 2, ptr nonnull %2, i8 0) #18
  call void @llvm.lifetime.end.p0(i64 8, ptr nonnull %_9.i)
  ret i64 %4
}
```
and `i686-unknown-uefi`:
```llvm-ir
; Function Attrs: noredzone nounwind
define i32 @efi_main(ptr %0, ptr %1) unnamed_addr #14 {
top:
  %_9.i = alloca ptr, align 4
  %2 = alloca [2 x ptr], align 4
  store ptr %0, ptr %2, align 4
  %3 = getelementptr inbounds ptr, ptr %2, i32 1
  store ptr %1, ptr %3, align 4
  call void @llvm.lifetime.start.p0(i64 4, ptr nonnull %_9.i)
  store ptr @_ZN11hello_world4main17hfe8e32f20eb88cc3E, ptr %_9.i, align 4
; call std::rt::lang_start_internal
  %4 = call i32 @_ZN3std2rt19lang_start_internal17he1faaadb6d46509dE(ptr noundef nonnull align 1 %_9.i, ptr noalias noundef nonnull readonly align 4 dereferenceable(12) @vtable.0, i32 2, ptr nonnull %2, i8 0) #18
  call void @llvm.lifetime.end.p0(i64 4, ptr nonnull %_9.i)
  ret i32 %4
}
```

As we can see, it's using [`Opaque Pointers`](https://llvm.org/docs/OpaquePointers.html).

Reply:
Yes, Opaque Pointers are the norm in recent LLVMs.

Reply:
Yes, Opaque Pointers are the norm in recent LLVMs.